metaflow/plugins/kfp/kfp.py

```diff
@@ -1,86 +1,275 @@
+    return step_op
+
+def pre_start_container_op():
+    """
+    Container op that corresponds to the 'pre-start' step of Metaflow.
+
+    Note: The public docker image is a copy of the internal docker image we were using (borrowed from aip-kfp-example).
+    """
+
+    pre_start_op = kfp.components.func_to_container_op(pre_start_op_func, base_image='ssreejith3/mf_on_kfp:python-curl-git')
+    return pre_start_op
+
+def create_command_templates_from_graph(graph):
+    """
+    Create a map of steps to their corresponding command templates. These command templates help define the command
+    to be used to run that particular step with placeholders for the `run_id` and `datastore_root` (location of the datastore).
+
+    # Note:
+    # Level-order traversal is adopted to keep the task-ids in line with what happens during a local metaflow execution.
+    # It is not entirely necessary to keep this order of task-ids if we are able to point to the correct input-paths for
+    # each step. But, using this ordering does keep the organization of data more understandable and natural (i.e., `start`
+    # gets a task id of 1, next step gets a task id of 2 and so on with 'end' step having the highest task id.
+    """
+    def build_cmd_template(step_name, task_id, input_paths):
+        python_cmd = "python {downloaded_file_name} --datastore s3 --datastore-root {{ds_root}} " \
+                     "step {step_name} --run-id {{run_id}} --task-id {task_id} " \
+                     "--input-paths {input_paths}".format(downloaded_file_name=DEFAULT_DOWNLOADED_FLOW_FILENAME,
+                                                            step_name=step_name, task_id=task_id, input_paths=input_paths)
+        return python_cmd
+
+    steps_deque = deque(['start']) # deque to process the DAG in level order
+    cur_task_id = 0
+
+    # set of seen steps, i.e., added to the queue for processing
+    seen_steps = set(['start'])
+    # Mapping of steps to task ids
+    task_id_map = {}
+    # Mapping of steps to their command templates
+    command_template_map = {}
+
+    while len(steps_deque) > 0:
+        cur_step = steps_deque.popleft()
+        cur_task_id += 1
+        task_id_map[cur_step] = cur_task_id
+
+        # Generate the correct input_path for each step. Note: input path depends on a step's parents (i.e., in_funcs)
+        # Format of the input-paths for reference:
+        # non-join nodes: run-id/parent-step/parent-task-id,
+        # branch-join node: run-id/:p1/p1-task-id,p2/p2-task-id,...
+        # foreach node: TODO: foreach is not considered here
+        if cur_task_id == 1: # start step
+            cur_input_path = '{run_id}/_parameters/0' # this is the standard input path for the `start` step
+        else:
+            if graph.nodes[cur_step].type != 'join':
+                parent_step = graph.nodes[cur_step].in_funcs[0]
```

Comment:
Can a `join` step have multiple `in_funcs`? If so shouldn't we be producing a list of the `cur_input_paths` to pass into `command_template_map`?

Looks like in the code sends through a `str` instead of a `List[str]`, looks like the code will still work but potentially would be nicer to be consistent.

Reply:
Yes, in this case it does have multiple `in-funcs` (branch joins have multiple in-funcs, foreach joins have just the one). 
So, the input path that is expected is a string and it looks different depending on the type of the node, but it's just one string. 

Branch join node: ` 1595911054530033/:bonus_movie/2,genre_movies/3 `
Foreach join node: `--input-paths 1595910350143453/compute_statistics/:2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23`
Linear node:  ` 1595911054530033/bonus_movie/2`